### PR TITLE
Write a document explaining Etcher's architecture

### PR DESCRIPTION
**Pull request by jviotti**
Signed-off-by: Juan Cruz Viotti <jviottidc@gmail.com>

### CODE REVIEW

File: docs/ARCHITECTURE.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+This document aims to serve as a high level overview of how Etcher works, specially oriented for contributors who want to understand the big picture.
+
+Module architecture
+-------------------
+
+Instead of embedding all the functionality required to create a full-featured image writer as a monolithic project, we try to hard to follow the ["lego block approach"](https://github.com/sindresorhus/ama/issues/10#issuecomment-117766328).
+
+This has the advantage of allowing other applications to re-use logic we implemented for Etcher in their own project, even for things we didn't expect, which leads to users benefitting from what we've built, and we benefitting from user's bug reports, suggestions, etc, as an indirect way to make Etcher better.
+
+The fact that low level details are scattered around many different modules makes it a bit hard for new contributor to wrap their heads around the project as a whole, and get a clear high level view of how things work or where to submit their work or bug reports.
```
Comment (craig-mulligan) [inline docs/ARCHITECTURE.md]: makes it a bit hard for new contributor -> can make it challenging for a new contributor

File: docs/ARCHITECTURE.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+As mentioned before, the Etcher GUI forks the CLI and retrieves information from it to update its state. In order to accomplish this, the Etcher CLI contains certain features to ease communication:
+
+- [Well-documented exit codes.](https://github.com/resin-io/etcher/blob/master/lib/src/exit-codes.js)
+- A `--robot` option, which causes the Etcher CLI to output state in a way that can be easily machine-parsed.
+- An `--ipc` option, which causes the Etcher CLI to use [NodeJS's built-in IPC channel](https://nodejs.org/api/child_process.html#child_process_child_send_message_sendhandle_options_callback) to emit messages from the child to the parent.
+
+GUI fifty-thousand foot view
+----------------------------
+
+Given the event oriented nature of desktop application, it can be hard to follow what's going on without getting deep in the details.
```
Comment (craig-mulligan) [inline docs/ARCHITECTURE.md]: nature of desktop application -> nature of desktop applications or nature of a desktop application.

File: docs/ARCHITECTURE.md (hunk 1 above)
Comment (craig-mulligan) [inline docs/ARCHITECTURE.md]: low level -> low-level

File: docs/ARCHITECTURE.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
+Etcher Architecture
+===================
+
+This document aims to serve as a high level overview of how Etcher works, specially oriented for contributors who want to understand the big picture.
```
Comment (craig-mulligan) [inline docs/ARCHITECTURE.md]: high-level

File: docs/ARCHITECTURE.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+Given the event oriented nature of desktop application, it can be hard to follow what's going on without getting deep in the details.
+
+To mitigate this, we try to encapsulate functionality with nice and straightforward interfaces as AngularJS modules, and provide a single place where all the modules are tied together.
+
+Therefore, if you want to get a rough idea of how the GUI works, the perfect place to start is [application controller](https://github.com/resin-io/etcher/blob/master/lib/gui/app.js) and the [main view](https://github.com/resin-io/etcher/blob/master/lib/gui/partials/main.html), and diving into specific modules depending on your interests.
+
+Summary
+-------
+
+We always welcome contributions to our documentation. If you want to give back to Etcher, but feel that your knowledge on how it works is not enough to tackle a bug report or feature requests, use that as your advantage, since your fresh eyes could help unveil things that we take for granted, but should be documented instead!
```
Comment (craig-mulligan) [inline docs/ARCHITECTURE.md]: your advantage, since your fresh eyes -> your advantage, since fresh eyes

Comment (craig-mulligan) [inline docs/ARCHITECTURE.md]: We always welcome contributions to our documentation -> We always welcome contributions to etcher as well as our documentation.

File: docs/ARCHITECTURE.md (hunk 3 above)
Comment (jviotti) [inline docs/ARCHITECTURE.md]: Fixed!

File: docs/ARCHITECTURE.md (hunk 1 above)
Comment (jviotti) [inline docs/ARCHITECTURE.md]: Both fixed!

File: docs/ARCHITECTURE.md (hunk 2 above)
Comment (jviotti) [inline docs/ARCHITECTURE.md]: Fixed!

File: docs/ARCHITECTURE.md (hunk 4 above)
Comment (jviotti) [inline docs/ARCHITECTURE.md]: Both fixed!
